test/router-luatest/map_part_test.lua

```diff
@@ -0,0 +1,299 @@
+    g.replica_1_a:exec(function()
+        ivshard.storage.internal.errinj.ERRINJ_BUCKET_GC_PAUSE = false
+        _G.bucket_gc_wait()
+    end)
+    -- Now move the buckets back to rs1 and pause gc on rs2.
+    -- The buckets will be ACTIVE on rs1 and SENT on rs2,
+    -- so the partial map should succeed.
+    res = g.replica_2_a:exec(function(bid1, bid2, to)
+        ivshard.storage.internal.errinj.ERRINJ_BUCKET_GC_PAUSE = true
+        ivshard.storage.bucket_send(bid1, to)
+        local res, err = ivshard.storage.bucket_send(bid2, to)
+        return {
+            res = res,
+            err = err,
+        }
+    end, {init.bids1[1], init.bids1[2], init.rs1_uuid})
+    t.assert_equals(res.err, nil)
+
+    res = g.router:exec(function(bid1, bid2)
+        local val, err, err_uuid = ivshard.router.map_part_callrw(
+            {bid1, bid2}, 'do_map', {42}, {timeout = iwait_timeout})
+        return {
+            val = val,
+            err = err,
+            err_uuid = err_uuid,
+        }
+    end, {init.bids1[1], init.bids1[2]})
+    t.assert_equals(res.err, nil)
+    t.assert_equals(res.err_uuid, nil)
+    expected = {
+        [init.rs1_uuid] = {{1, 42}},
+    }
+    t.assert_equals(res.val, expected)
+    -- Re-enable gc on rs2.
+    g.replica_2_a:exec(function()
+        ivshard.storage.internal.errinj.ERRINJ_BUCKET_GC_PAUSE = false
+        _G.bucket_gc_wait()
+    end)
+end
+
+g.test_map_part_double_ref = function(g)
+    local expected, res
+    local init = map_part_init()
+
+    -- First, disable discovery on the router to disable route cache update.
+    g.router:exec(function()
+        ivshard.router.internal.errinj.ERRINJ_LONG_DISCOVERY = true
+    end)
+    -- Then, move the bucket form rs1 to rs2. Now the router has an outdated
+    -- route cache.
+    g.replica_1_a:exec(function(bid, to)
+        ivshard.storage.bucket_send(bid, to)
+    end, {init.bids1[4], init.rs2_uuid})
+    -- Call a partial map for the moved bucket and some bucket from rs2. The ref stage
+    -- should be done in two steps:
+    -- 1. ref rs2 and returns the moved bucket;
+    -- 2. discover the moved bucket on rs2 and avoid double reference;
+    res = g.router:exec(function(bid1, bid2)
+        local val, err, err_uuid = ivshard.router.map_part_callrw(
+            {bid1, bid2}, 'do_map', {42}, {timeout = iwait_timeout})
+        return {
+            val = val,
+            err = err,
+            err_uuid = err_uuid,
+        }
+    end, {init.bids1[4], init.bids2[1]})
+    t.assert_equals(res.err, nil)
+    t.assert_equals(res.err_uuid, nil)
+    expected = {
+        [init.rs2_uuid] = {{2, 42}},
+    }
+    t.assert_equals(res.val, expected)
+    -- Call a partial map one more time to make sure there are no references left.
+    res = g.router:exec(function(bid1, bid2)
+        local val, err, err_uuid = ivshard.router.map_part_callrw(
+            {bid1, bid2}, 'do_map', {42}, {timeout = iwait_timeout})
```

Comment:
How does calling `map_part_callrw()` second time ensures that there are no references left? It will succeed even if there are references anyway.